fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteToVectorPlanRule.java

```diff
@@ -0,0 +1,231 @@
+
+        Map<ColumnRefOperator, ScalarOperator> projectMap = new HashMap<>();
+        scanOperator.getProjection().getColumnRefMap().entrySet().stream()
+                .forEach(entry -> {
+                    if (FunctionSet.VECTOR_COMPUTE_FUNCTIONS.contains(entry.getKey().getName())
+                            && entry.getValue() instanceof CallOperator) {
+                        projectMap.put(distanceColumnRefOperator, distanceColumnRefOperator);
+                    } else {
+                        projectMap.put(entry.getKey(), entry.getValue());
+                    }
+                });
+        newScanOperator.setProjection(new Projection(projectMap));
+
+        List<Ordering> orderByElements = topNOperator.getOrderByElements().stream().map(ordering ->
+                FunctionSet.VECTOR_COMPUTE_FUNCTIONS.contains(ordering.getColumnRef().getName()) ?
+                    new Ordering(distanceColumnRefOperator, ordering.isAscending(), ordering.isNullsFirst()) : ordering)
+                .collect(Collectors.toList());
+
+        Map<ColumnRefOperator, ScalarOperator> topNProjectMap = new HashMap<>();
+        boolean hasProjection = topNOperator.getProjection() != null;
+        if (hasProjection) {
+            topNOperator.getProjection().getColumnRefMap().entrySet().stream()
+                    .forEach(entry -> {
+                        if (FunctionSet.VECTOR_COMPUTE_FUNCTIONS.contains(entry.getKey().getName())) {
+                            topNProjectMap.put(distanceColumnRefOperator, distanceColumnRefOperator);
+                        } else {
+                            topNProjectMap.put(entry.getKey(), entry.getValue());
+                        }
+                    });
+        }
+
+        // new TopN operator
+        LogicalTopNOperator newTopNOperator = new LogicalTopNOperator(topNOperator.getLimit(),
+                topNOperator.getPredicate(), hasProjection ? new Projection(topNProjectMap) : null,
+                topNOperator.getPartitionByColumns(), topNOperator.getPartitionLimit(), orderByElements,
+                topNOperator.getOffset(), topNOperator.getSortPhase(), topNOperator.getTopNType(), topNOperator.isSplit());
+
+        OptExpression topNExpression = OptExpression.create(newTopNOperator);
+        topNExpression.getInputs().clear();
+        topNExpression.getInputs().add(OptExpression.create(newScanOperator));
+
+        return topNExpression;
+    }
+
+    // todo: prune in dfs
+    public Optional<ScalarOperator> findAndSetVectorRange(ScalarOperator operator,
+                SessionVariable sessionVariable, boolean isAscending) {
+        if (!sessionVariable.isEnableUseANN()) {
+            return Optional.empty();
+        }
+
+        if (operator instanceof BinaryPredicateOperator && operator.getChild(1) instanceof ConstantOperator &&
+                (operator.getChild(0).toString().contains("l2") || operator.getChild(0).toString().contains("cosine"))) {
```

Comment:
use string contains method is too tricky. Column alias can be anything, it may match this condition.